src/libmeasurement_kit/dns/resolver.hpp

```diff
@@ -0,0 +1,152 @@
+        delete ctx;
+        return;
+    }
+
+    if (ctx->servinfo == nullptr) {
+        ctx->reactor->call_soon(
+            [callback]() { callback(ResolverError(), nullptr); });
+        delete ctx;
+        return;
+    }
+
+    void *addr_ptr;
+    char address[128];
+    std::vector<Answer> answers;
+    for (p = ctx->servinfo; p != nullptr; p = p->ai_next) {
+        Answer answer;
+        answer.name = ctx->name;
+        answer.qclass = ctx->dns_class;
+        if (p->ai_family == AF_INET) {
+            answer.type = QueryTypeId::A;
+            addr_ptr = &((struct sockaddr_in *)p->ai_addr)->sin_addr;
+        } else if (p->ai_family == AF_INET6) {
+            answer.type = QueryTypeId::AAAA;
+            addr_ptr = &((struct sockaddr_in6 *)p->ai_addr)->sin6_addr;
+        } else {
+            ctx->reactor->call_soon(
+                [callback]() { callback(ResolverError(), nullptr); });
+            delete ctx;
+            return;
+        }
+        if (inet_ntop(p->ai_family, addr_ptr, address, sizeof(address)) ==
+            NULL) {
+            ctx->logger->warn("dns: unexpected inet_ntop failure");
+            throw std::runtime_error("Unexpected inet_ntop failure");
+        }
+        if (p->ai_family == AF_INET) {
+            answer.ipv4 = std::string(address);
+        } else if (p->ai_family == AF_INET6) {
+            answer.ipv6 = std::string(address);
+        }
+        answers.push_back(answer);
+    }
+    ctx->message->answers = answers;
+    message_copy = ctx->message;
+    ctx->reactor->call_soon(
+        [callback, message_copy]() { callback(NoError(), message_copy); });
+    delete ctx;
+}
+
+inline void system_resolver(QueryClass dns_class, QueryType dns_type,
+                            std::string name, Callback<Error, Var<Message>> cb,
+                            Settings settings, Var<Reactor> reactor,
+                            Var<Logger> logger) {
+    ResolverContext *ctx = new ResolverContext(dns_class, dns_type, name, cb,
+                                               settings, reactor, logger);
+    Query query;
+    ctx->hints.ai_flags = AI_ADDRCONFIG;
+    ctx->hints.ai_socktype = SOCK_STREAM;
+    ctx->hints.ai_protocol = 0;
+    ctx->hints.ai_canonname = NULL;
+    ctx->hints.ai_addr = NULL;
+    ctx->hints.ai_next = NULL;
+
+    if (dns_class != QueryClassId::IN) {
+        cb(UnsupportedClassError(), nullptr);
+        return;
+    }
+
+    if (dns_type == QueryTypeId::A) {
+        ctx->hints.ai_family = AF_UNSPEC;
+    } else if (dns_type == QueryTypeId::AAAA) {
+        ctx->hints.ai_family = AF_INET6;
+    } else {
+        cb(UnsupportedTypeError(), nullptr);
+        return;
+    }
+
+    query.type = dns_type;
+    query.qclass = dns_class;
+    query.name = name;
+
+    ctx->message->queries.push_back(query);
+
+    std::thread res_thread(resolve_async, ctx);
+    /// XXX study a way to mantain a reference to the thread
+    res_thread.detach();
```

Comment:
Some thoughts on the code in here:

1) I'd rather use `std::launch::async()` because it's higher level than `std::thread` and I think it's better to use the highest layer API that does the job for us

2) [however?] I imagine in the future we will have a queue for running "tasks" like this one, because it's not super robust to possible spawn N threads to do N concurrent DNS resolutions... I'm not saying we should be doing this right now but perhaps we should take this possibility into account

3) I'd rather use a lambda here instead of creating a separate function (but I guess this is a matter of personal taste and one could also argue against using a lambda... yet, probably when we have a task queue one simple way to implement it would be "call this lambda in a separate thread and callback in the main thread when done")